File: pkg/dart2cpp/sample/cpp_generated/test_is_type_promotion_ternary.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 类: Animal
// ============================================================================

class Animal {
public:
  String name;
  Animal(String name) : name(name) {
  }
  
};

// ============================================================================
// 类: Dog
// ============================================================================

class Dog : public Animal {
public:
  String breed;
  Dog(String name, String breed) : breed(breed), Animal(name) {
  }
  
  Nullable bark() {
    dart_print((this->name).toString() + dart_string(" says: Woof! Woof!"));
return Void;
  }
  
};

// ============================================================================
// 类: Cat
// ============================================================================

class Cat : public Animal {
public:
  Bool isLazy;
  Cat(String name, Bool isLazy) : isLazy(isLazy), Animal(name) {
  }
  
  Nullable meow() {
    dart_print((this->name).toString() + dart_string(" says: Meow!"));
return Void;
  }
  
};

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    auto animal1 = ObjectPtr<Dog>(new Dog(dart_string("Rex"), dart_string("Labrador")));
auto result1 = dart_is<ObjectPtr<Dog>>(animal1) ? dart_cast<ObjectPtr<Dog>>(animal1)->bark() : dart_print(dart_string("Not a dog"));
auto animal2 = ObjectPtr<Cat>(new Cat(dart_string("Whiskers"), dart_bool(true)));
auto result2 = dart_is<ObjectPtr<Cat>>(animal2) && dart_cast<ObjectPtr<Cat>>(animal2)->isLazy ? dart_cast<ObjectPtr<Cat>>(animal2)->meow() : dart_print(dart_string("Not a lazy cat"));
auto obj1 = dart_string("Hello");
auto obj2 = dart_int(42);
auto result3 = dart_is<String>(obj1) && dart_is<Int>(obj2) ? dart_print(dart_concat((obj1->size()).toString(), dart_string(" and "), (obj2).toString())) : dart_print(dart_string("Type mismatch"));
auto animal3 = ObjectPtr<Dog>(new Dog(dart_string("Buddy"), dart_string("Golden Retriever")));
auto result4 = dart_is<ObjectPtr<Dog>>(animal3) && dart_cast<ObjectPtr<Dog>>(animal3)->breed->length->>(dart_int(5)) && dart_cast<ObjectPtr<Dog>>(animal3)->name->length->>(dart_int(3)) ? dart_print(dart_string("Long breed name: ") + (animal3->breed).toString()) : dart_print(dart_string("Short name"));
auto animal4 = ObjectPtr<Cat>(new Cat(dart_string("Tom"), dart_bool(false)));
auto result5 = dart_is<ObjectPtr<Cat>>(animal4) ? animal4->isLazy ? dart_print(dart_string("Lazy cat")) : dart_cast<ObjectPtr<Cat>>(animal4)->meow() : dart_print(dart_string("Not a cat"));
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}

File: pkg/dart2cpp/sample/dart/test_is_type_promotion_ternary.dart
// 测试三元表达式中的 is 类型推断和类型提升

class Animal {
  String name;
  Animal(this.name);
}

class Dog extends Animal {
  String breed;

  Dog(String name, this.breed) : super(name);

  void bark() {
    print('$name says: Woof! Woof!');
  }
}

class Cat extends Animal {
  bool isLazy;

  Cat(String name, this.isLazy) : super(name);

  void meow() {
    print('$name says: Meow!');
  }
}

void main() {
  // 测试1: 简单的 is 检查在三元表达式条件中
  Animal animal1 = Dog('Rex', 'Labrador');
  var result1 = animal1 is Dog ? animal1.bark() : print('Not a dog');

  // 测试2: is 检查 && 另一个条件
  Animal animal2 = Cat('Whiskers', true);
  var result2 = animal2 is Cat && animal2.isLazy
      ? animal2.meow()
      : print('Not a lazy cat');

  // 测试3: 多个 is 检查用 && 连接
  var obj1 = 'Hello';
  var obj2 = 42;
  var result3 = obj1 is String && obj2 is int
      ? print('${obj1.length} and $obj2')
      : print('Type mismatch');

  // 测试4: 复杂的 && 链式条件
  Animal animal3 = Dog('Buddy', 'Golden Retriever');
  var result4 =
      animal3 is Dog && animal3.breed.length > 5 && animal3.name.length > 3
          ? print('Long breed name: ${animal3.breed}')
          : print('Short name');

  // 测试5: 嵌套三元表达式中的 is 检查
  Animal animal4 = Cat('Tom', false);
  var result5 = animal4 is Cat
      ? (animal4.isLazy ? print('Lazy cat') : animal4.meow())
      : print('Not a cat');
}
